### PR TITLE
Rework checkpointing to avoid race conditions and incorrectly checkpointed exceptions

### PR DESCRIPTION
**Pull request by benclifford**
Although mostly this doesn't happen, in the case of issue #2252,
a KeyError was being raised by hash_lookup, and this was being
placed into the checkpoint result rather than propagated.

This PR reworks that code to not use hash_lookup at all, and
remove that entire exception code path.

This does not address #2252 but makes the behaviour a bit more
consistent.

## Type of change

- Bug fix (non-breaking change that fixes an issue)

### CODE REVIEW

File: parsl/dataflow/dflow.py [hunk 1]
```diff
@@ -1158,15 +1158,11 @@ def checkpoint(self, tasks=None):
...
-                            # TODO : Resolve this question ?
-                            r = self.memoizer.hash_lookup(hashsum).result()
-                        except Exception as e:
-                            t['exception'] = e
-                        else:
-                            t['result'] = r
+
+                        fut = self.memoizer.hash_lookup(hashsum)
+                        assert fut.done()
+                        assert fut.exception() is None
```
Comment (yadudoc) [inline parsl/dataflow/dflow.py:1164]: Aren't we already checking this in the condition above on line 1152?

**Review by yadudoc — COMMENTED** [review]
I agree that we shouldn't checkpoint exceptions. It looks like there's checks on that already in the code, so I'm not sure what the assertion adds.

File: parsl/dataflow/dflow.py (hunk 1 above)
Comment (benclifford) [reply parsl/dataflow/dflow.py:1164]: The test on line 1152 checks  self.tasks[task_id]['app_fu'], not self.memoizer.hash_lookup(hashsum)

Is the result of the hash_lookup meant to be the same as self.tasks[task_id]['app_fu'] ? I think probably so?
 
In which case that hash_lookup call is probably unnecessary?

Comment (benclifford) [reply parsl/dataflow/dflow.py:1164]: (i.e. the task record should be the source of truth here, not the hash lookup table?)

**Comment by benclifford** [comment]
I rearranged this code to use the task record as source of truth, instead of expecting the value to have been placed in the memo lookup table. This should eliminate the race condition in #2252, which my first commits in this PR did not do.

**Comment by benclifford** [comment]
I still need to test that against my race-condition-provoking patch in #2252, though

**Comment by benclifford** [comment]
I have validated this PR against the reproducible failure case in #2252

**Pull request merged**